internal/v2/tls_config_store/tls_config_store.go

```diff
@@ -22,53 +24,188 @@ var (
 )
 
 // GetTlsConfigurationForClient returns a tls.Config instance for use by a client application.
-func GetTlsConfigurationForClient(serverHostname string) *tls.Config {
-	// TODO(rmehta19): Call S2Av2 for certificate.
-	// TODO(rmehta19): Call remote signer library for private key.
-	cert, err := tls.X509KeyPair(clientCert, clientKey)
+func GetTlsConfigurationForClient(serverHostname string, cstream s2av2pb.S2AService_SetUpSessionClient) *tls.Config {
+	// Send request to S2Av2 for config.
+	err := cstream.Send(&s2av2pb.SessionReq {
+		AuthenticationMechanisms: []*s2av2pb.AuthenticationMechanism {
+			{
+				// TODO(rmehta19): Populate Authentication Mechanism using tokenmanager.
+				MechanismOneof: &s2av2pb.AuthenticationMechanism_Token{"token"},
+			},
+		},
+		ReqOneof: &s2av2pb.SessionReq_GetTlsConfigurationReq {
+			&s2av2pb.GetTlsConfigurationReq {
+				ConnectionSide: commonpb.ConnectionSide_CONNECTION_SIDE_CLIENT,
+			},
+		},
+	})
 	if err != nil {
-		log.Fatalf("Failed to generate X509KeyPair: %v", err)
+		log.Fatalf("Client: failed to send SessionReq: %v", err)
+	}
+	log.Printf("Client: sent SessionReq")
+
+	// Get the response containing config from S2Av2.
+	resp, err := cstream.Recv()
+	if err != nil {
+		log.Fatalf("Client: failed to recieve SessionResp: %v", err)
+	}
+	log.Printf("Client: recieved SessionResp")
+
+	// TODO(rmehta19): Handle resp.GetStatus().
+
+	// Extract TLS configiguration from SessionResp.
+	tlsConfig := resp.GetGetTlsConfigurationResp().GetClientTlsConfiguration()
+
+	// Fake S2Av2 only puts one cert in CertitificateChain, for loop iterates
+	// once.
+	var certList []tls.Certificate
+	for _, v := range tlsConfig.CertificateChain {
+		// TODO(rmehta19): Call remote signer library for private key.
+		cert, err := tls.X509KeyPair([]byte(v), clientKey)
```

Comment:
I see, I used a for loop here, because in the future we might have a chain of certs from leaf to root. For the fake implementation of the s2av2 service layer, the chain is only of length 1, which is why I hardcoded the key. I thought that in the future, in the for loop, we can make a call to the remote signer library for the private key corresponding to a cert(remote signer lib then calls s2av2 for this info) from the list.

Does s2av2 only maintain private key for the leaf cert in the certificate chain?

Reply:
Also, another question, with these two if statements, if I'm understanding correctly, we will end up creating a cert with an empty (PrivateKey field)[https://pkg.go.dev/crypto/tls#Certificate]. Is it ok to populate the config with certs that don't have that field populated?

Reply:
Correct that the `GetTlsConfigurationResp` will, in general, consist of one or more certificates and that the S2A only maintains the private key for the leaf certificate.

RE the second point: this is my bad, I misunderstood how the API works. Apologies!

Let me expand on how I think this should work:

* We need to set the `Certificates` field of the `tls.Config`. We should populate it with an array containing exactly 1 instance of the `Certificate` struct.
* To create this instance of the `Certificate` struct, we need to populate the `Certificates` field and the `PrivateKey` field.
* To populate the `Certificates` field, we do the following: for each cert in the `GetTlsConfigurationResp`, we convert it from a PEM string to a DER byte array, and add this array to the `Certificates` field. I think this conversion can be done using the `pem.Decode` function, based on this documentation: https://pkg.go.dev/encoding/pem#Decode
* To populate the `PrivateKey` field, we should only use the private key corresponding to the leaf cert (in the short term), which we will later replace with the RemoteSigner.

Does that make sense slash answer the question?

Reply:
Thanks for the detailed explanation Matt! Yes, this definitely clears up the problem I was having. I understood the S2Av2 gives us the certificate chain for the client / server, what I then (incorrectly) did is use the for loop to generate a new tls.Certificate for each cert in the chain, I then appended each certificate to list of []tls.Certificate, to pass to the config. After this explanation and today's standup, I understand that each cert in the chain S2Av2 gives us, will be parsed and appended to a singular tls.Certificate. 

The certificate decoding and parsing works using the pem.Decode and x509.Parsecertificate, the issue I faced after that was populating the PrivateKey field of the tls.Certificate. Since the Remote Signer fake is not yet implemented, I found this workaround:

Since the fake supplies the cert chain from leaf to root, for both the server and the client, I call:
`tlsCert, err := tls.X509KeyPair([]byte(tlsConfig.CertificateChain[0]), client/serverKey)`
`cert.PrivateKey = tlsCert.PrivateKey`
where `cert` is the tls.Certificate object we appended to in the for loop where we iterated through the S2Av2 provided cert chain. tlsCert is essentialy an intermediate variable, it serves no purpose other than the fact that I use its PrivateKey field for my cert object. I plan on removing this workaround after the Private Key operation fake is implented and called from the remote signer lib(which implements the crypto.Signer interface), which will be called from here. It's not the most elegant thing to generate a TLS cert only to use its PrivateKey field, but right now, I think it's the simplest. Does this sound ok? Do you think there's a better way to go about this?

Reply:
Thanks for the detailed explanation Matt! Yes, this definitely clears up the problem I was having. I understood the S2Av2 gives us the certificate chain for the client / server, what I then (incorrectly) did is use the for loop to generate a new tls.Certificate for each cert in the chain, I then appended each certificate to list of []tls.Certificate, to pass to the config. After this explanation and today's standup, I understand that each cert in the chain S2Av2 gives us, will be parsed and appended to a singular tls.Certificate. 

The certificate decoding and parsing works using the pem.Decode and x509.Parsecertificate, the issue I faced after that was populating the PrivateKey field of the tls.Certificate. Since the Remote Signer fake is not yet implemented, I found this workaround:

Since the fake supplies the cert chain from leaf to root, for both the server and the client, I call:
`tlsCert, err := tls.X509KeyPair([]byte(tlsConfig.CertificateChain[0]), client/serverKey)`
`cert.PrivateKey = tlsCert.PrivateKey`
where `cert` is the tls.Certificate object we appended to in the for loop where we iterated through the S2Av2 provided cert chain. tlsCert is essentialy an intermediate variable, it serves no purpose other than the fact that I use its PrivateKey field for my cert object. I plan on removing this workaround after the Private Key operation fake is implented and called from the remote signer lib(which implements the crypto.Signer interface), which will be called from here. It's not the most elegant thing to generate a TLS cert only to use its PrivateKey field, but right now, I think it's the simplest. Does this sound ok? Do you think there's a better way to go about this?

Reply:
Discussed offline, this is all good, added todo to move this private key stuff to remote signer.